File: src/CPort/instructions.c
#include "emulator_port.h"

#include <stdio.h>

int NUL(mother_board_t *mother_board, void *dev)
{
    return (0x00);
}

int ICX(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->x += 1;
    return (0x00);
}

int ICY(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->y += 1;
    return (0x00);
}

int ADX(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->x += casted_dev->registries->temp;
    return (0x00);
}

int ADY(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->y += casted_dev->registries->temp;
    return (0x00);
}

int SUX(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->x -= casted_dev->registries->temp;
    return (0x00);
}

int SUY(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->y -= casted_dev->registries->temp;
    return (0x00);
}

int MUX(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->x *= casted_dev->registries->temp;
    return (0x00);
}

int MUY(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->y *= casted_dev->registries->temp;
    return (0x00);
}

int DIX(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->x /= casted_dev->registries->temp;
    return (0x00);
}

int DIY(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->x /= casted_dev->registries->temp;
    return (0x00);
}

int MOX(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->x %= casted_dev->registries->temp;
    return (0x00);
}

int MOY(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->y %= casted_dev->registries->temp;
    return (0x00);
}


int TVF(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->temp = casted_dev->registries->fetched;
    return (0x00);
}

int END(mother_board_t *mother_board, void *dev)
{
    return (0x00);
}

int TVX(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->temp = casted_dev->registries->x;
    return (0x00);
}

int TVY(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->temp = casted_dev->registries->y;
    return (0x00);
}

int PPX(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->ptr_prg = casted_dev->registries->x;
    casted_dev->registries->ptr_prg -= 1;
    return (0x00);
}

int PPY(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->ptr_prg = casted_dev->registries->y;
    casted_dev->registries->ptr_prg -= 1;
    return (0x00);
}

int MPX(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->ptr_mem = casted_dev->registries->x;
    return (0x00);
}

int MPY(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->ptr_mem = casted_dev->registries->y;
    return (0x00);
}

int DMX(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->mem_device = casted_dev->registries->x;
    return (0x00);
}

int DMY(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->mem_device = casted_dev->registries->y;
    return (0x00);
}

int DPX(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->prg_device = casted_dev->registries->x;
    return (0x00);
}

int DPY(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->prg_device = casted_dev->registries->y;
    return (0x00);
}

int XVY(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->x = casted_dev->registries->y;
    return (0x00);
}

int YVX(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->y = casted_dev->registries->x;
    return (0x00);
}

int XVT(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->x = casted_dev->registries->temp;
    return (0x00);
}

int YVT(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->y = casted_dev->registries->temp;
    return (0x00);
}

int XVF(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->x = casted_dev->registries->fetched;
    return (0x00);
}

int YVF(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->y = casted_dev->registries->fetched;
    return (0x00);
}

int MVX(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    if (mother_board->ram->id == casted_dev->registries->mem_device)
        mother_board->ram->write(casted_dev->registries->ptr_mem, casted_dev->registries->x, mother_board->ram);
    return (0x00);
}

int MVY(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    if (mother_board->ram->id == casted_dev->registries->mem_device)
        mother_board->ram->write(casted_dev->registries->ptr_mem, casted_dev->registries->y, mother_board->ram);
    return (0x00);
}

int RDM(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    if (mother_board->ram->id == casted_dev->registries->mem_device)
        casted_dev->registries->fetched = mother_board->ram->read(casted_dev->registries->ptr_mem, mother_board->ram);
    return (0x00);
}

int TMX(mother_board_t *mother_board, void *dev)
{
    printf("TMX not implemented.\n");
    return (0x00);
}

int TMY(mother_board_t *mother_board, void *dev)
{
    printf("TMX not implemented.\n");
    return (0x00);
}

int ANX(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->x &= casted_dev->registries->temp;
    return (0x00);
}

int ANY(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->y &= casted_dev->registries->temp;
    return (0x00);
}

int ORX(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->x |= casted_dev->registries->temp;
    return (0x00);
}

int ORY(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->y |= casted_dev->registries->temp;
    return (0x00);
}

int XOX(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->x ^= casted_dev->registries->temp;
    return (0x00);
}

int XOY(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->y ^= casted_dev->registries->temp;
    return (0x00);
}

int LSX(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->x <<= casted_dev->registries->temp;
    return (0x00);
}

int LSY(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->y <<= casted_dev->registries->temp;
    return (0x00);
}

int RSX(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->x >>= casted_dev->registries->temp;
    return (0x00);
}

int RSY(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->y >>= casted_dev->registries->temp;
    return (0x00);
}

int EQX(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->x = casted_dev->registries->x == casted_dev->registries->temp;
    return (0x00);
}

int EQY(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->y = casted_dev->registries->y == casted_dev->registries->temp;
    return (0x00);
}

int NQX(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->x = casted_dev->registries->x != casted_dev->registries->temp;
    return (0x00);
}

int NQY(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->y = casted_dev->registries->y != casted_dev->registries->temp;
    return (0x00);
}

int SPX(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->x = casted_dev->registries->x > casted_dev->registries->temp;
    return (0x00);
}

int SPY(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->y = casted_dev->registries->y > casted_dev->registries->temp;
    return (0x00);
}

int INX(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->x = casted_dev->registries->x < casted_dev->registries->temp;
    return (0x00);
}

int INY(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->y = casted_dev->registries->y < casted_dev->registries->temp;
    return (0x00);
}

int IEX(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->x = casted_dev->registries->x <= casted_dev->registries->temp;
    return (0x00);
}

int IEY(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->y = casted_dev->registries->y <= casted_dev->registries->temp;
    return (0x00);
}

int SEX(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->x = casted_dev->registries->x >= casted_dev->registries->temp;
    return (0x00);
}

int SEY(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->y = casted_dev->registries->y >= casted_dev->registries->temp;
    return (0x00);
}

int NOX(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->x = ~casted_dev->registries->x;
    return (0x00);
}

int NOY(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->y = ~casted_dev->registries->y;
    return (0x00);
}

int CNE(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    if (!casted_dev->registries->temp)
        casted_dev->registries->ptr_prg += 1;
    return (0x00);
}

int HALT(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->ptr_prg -= 1;
    return (0x00);
}

int XPP(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->x = casted_dev->registries->ptr_prg;
    return (0x00);
}

int YPP(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->y = casted_dev->registries->ptr_prg;
    return (0x00);
}

int SWX(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;
    int temp;

    temp = casted_dev->registries->temp;
    casted_dev->registries->temp = casted_dev->registries->x;
    casted_dev->registries->x = temp;
    return (0x00);
}

int SWY(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;
    int temp;

    temp = casted_dev->registries->temp;
    casted_dev->registries->temp = casted_dev->registries->y;
    casted_dev->registries->y = temp;
    return (0x00);
}

int REX(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->x = 0x00;
    return (0x00);
}

int REY(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->y = 0x00;
    return (0x00);
}

int XMP(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->x = casted_dev->registries->ptr_mem;
    return (0x00);
}

int YMP(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->y = casted_dev->registries->ptr_mem;
    return (0x00);
}

int STX(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->ptr_prg += 1;
    if (mother_board->ram->id == casted_dev->registries->prg_device)
        casted_dev->registries->x = mother_board->ram->read(casted_dev->registries->ptr_prg, mother_board->ram);
    return (0x00);
}

int STY(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    casted_dev->registries->ptr_prg += 1;
    if (mother_board->ram->id == casted_dev->registries->prg_device)
        casted_dev->registries->y = mother_board->ram->read(casted_dev->registries->ptr_prg, mother_board->ram);
    return (0x00);
}

int LOG(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    printf("%c", casted_dev->registries->temp);
    return (0x00);
}

int IDN(mother_board_t *mother_board, void *dev)
{
    cpu_t *casted_dev = (cpu_t *)dev;

    return (casted_dev->id);
}

File: src/JSPort/instructions.js
const { get_device } = require("./common.js");

module.exports = {
    NULL: function (motherboard, self) {
        return (0x00);
    },

    ICX: function (motherboard, self) {
        self.registries["x"] += 1;
        return (0x00);
    },

    ICY: function (motherboard, self) {
        self.registries["y"] += 1;
        return (0x00);
    },

    ADX: function (motherboard, self) {
        self.registries["x"] += self.registries["temp"];
        return (0x00);
    },

    ADY: function (motherboard, self) {
        self.registries["y"] += self.registries["temp"];
        return (0x00);
    },

    SUX: function (motherboard, self) {
        self.registries["x"] -= self.registries["temp"];
        return (0x00);
    },

    IDN: function (motherboard, self) {
        return (self.id);
    },

    SUY: function (motherboard, self) {
        self.registries["y"] -= self.registries["temp"];
        return (0x00);
    },

    MUX: function (motherboard, self) {
        self.registries["x"] *= self.registries["temp"];
        return (0x00);
    },

    MUY: function (motherboard, self) {
        self.registries["y"] *= self.registries["temp"];
        return (0x00);
    },

    DIX: function (motherboard, self) {
        self.registries["x"] = Math.round(self.registries["x"] / self.registries["temp"]);
        return (0x00);
    },

    DIY: function (motherboard, self) {
        self.registries["y"] = Math.round(self.registries["y"] / self.registries["temp"]);
        return (0x00);
    },

    MOX: function (motherboard, self) {
        self.registries["x"] %= self.registries["temp"];
        return (0x00);
    },

    MOY: function (motherboard, self) {
        self.registries["y"] %= self.registries["temp"];
        return (0x00);
    },

    TVF: function (motherboard, self) {
        self.registries["temp"] = self.registries["fetched"];
        return (0x00);
    },

    END: function (motherboard, self) {
        return (0x00);
    },

    TVX: function (motherboard, self) {
        self.registries["temp"] = self.registries["x"];
        return (0x00);
    },

    TVY: function (motherboard, self) {
        self.registries["temp"] = self.registries["y"];
        return (0x00);
    },

    PPX: function (motherboard, self) {
        self.registries["ptr_prg"] = self.registries["x"];
        self.registries["ptr_prg"] -= 1;
        return (0x00);
    },

    PPY: function (motherboard, self) {
        self.registries["ptr_prg"] = self.registries["y"];
        self.registries["ptr_prg"] -= 1;
        return (0x00);
    },

    MPX: function (motherboard, self) {
        self.registries["ptr_mem"] = self.registries["x"];
        return (0x00);
    },

    MPY: function (motherboard, self) {
        self.registries["ptr_mem"] = self.registries["y"];
        return (0x00);
    },

    DMX: function (motherboard, self) {
        self.registries["mem_device"] = self.registries["x"];
        return (0x00);
    },

    DMY: function (motherboard, self) {
        self.registries["mem_device"] = self.registries["y"];
        return (0x00);
    },

    DPX: function (motherboard, self) {
        self.registries["mem_device"] = self.registries["x"];
        return (0x00);
    },

    DPY: function (motherboard, self) {
        self.registries["mem_device"] = self.registries["y"];
        return (0x00);
    },

    XVY: function (motherboard, self) {
        self.registries["x"] = self.registries["y"];
        return (0x00);
    },

    YVX: function (motherboard, self) {
        self.registries["y"] = self.registries["x"];
        return (0x00);
    },

    XVT: function (motherboard, self) {
        self.registries["x"] = self.registries["temp"];
        return (0x00);
    },

    YVT: function (motherboard, self) {
        self.registries["y"] = self.registries["temp"];
        return (0x00);
    },

    XVF: function (motherboard, self) {
        self.registries["x"] = self.registries["fetched"];
        return (0x00);
    },

    YVF: function (motherboard, self) {
        self.registries["y"] = self.registries["fetched"];
        return (0x00);
    },

    MVX: function (motherboard, self) {
        var dev = get_device(motherboard, self.registries["mem_device"]);

        if (dev == undefined)
            return (0x01);
        dev.write(self.registries["ptr_mem"], self.registries["x"]);
        return (0x00);
    },

    MVY: function (motherboard, self) {
        var dev = get_device(motherboard, self.registries["mem_device"]);

        if (dev == undefined)
            return (0x01);
        dev.write(self.registries["ptr_mem"], self.registries["y"]);
        return (0x00);
    },

    RDM: function (motherboard, self) {
        var dev = get_device(motherboard, self.registries["mem_device"]);

        if (dev == undefined)
            return (0x01);
            self.registries["fetched"] = dev.read(self.registries["ptr_mem"]);
        return (0x00);
    },

    TMX: function (motherboard, self) {
        var dev = get_device(motherboard, self.registries["temp"]);

        if (dev == undefined)
            return (0x01);
        self.registries["fetched"] = dev.send_message(motherboard, self.registries["x"]);
        return (0x00);
    },

    TMY: function (motherboard, self) {
        var dev = get_device(motherboard, self.registries["temp"]);

        if (dev == undefined)
            return (0x01);
        self.registries["fetched"] = dev.send_message(motherboard, self.registries["y"]);
        return (0x00);
    },

    ANX: function (motherboard, self) {
        self.registries["x"] &= self.registries["temp"];
        return (0x00);
    },

    ANY: function (motherboard, self) {
        self.registries["y"] &= self.registries["temp"];
        return (0x00);
    },

    ORX: function (motherboard, self) {
        self.registries["x"] |= self.registries["temp"];
        return (0x00);
    },

    ORY: function (motherboard, self) {
        self.registries["y"] |= self.registries["temp"];
        return (0x00);
    },

    XOX: function (motherboard, self) {
        self.registries["x"] ^= self.registries["temp"];
        return (0x00);
    },

    XOY: function (motherboard, self) {
        self.registries["y"] ^= self.registries["temp"];
        return (0x00);
    },

    LSX: function (motherboard, self) {
        self.registries["x"] <<= self.registries["temp"];
        return (0x00);
    },

    LSY: function (motherboard, self) {
        self.registries["y"] <<= self.registries["temp"];
        return (0x00);
    },

    RSX: function (motherboard, self) {
        self.registries["x"] >>= self.registries["temp"];
        return (0x00);
    },

    RSY: function (motherboard, self) {
        self.registries["y"] >>= self.registries["temp"];
        return (0x00);
    },

    EQX: function (motherboard, self) {
        self.registries["x"] = self.registries["x"] == self.registries["temp"];
        return (0x00);
    },

    EQY: function (motherboard, self) {
        self.registries["y"] = self.registries["y"] == self.registries["temp"];
        return (0x00);
    },

    NQX: function (motherboard, self) {
        self.registries["x"] = self.registries["x"] != self.registries["temp"];
        return (0x00);
    },

    NQY: function (motherboard, self) {
        self.registries["y"] = self.registries["y"] != self.registries["temp"];
        return (0x00);
    },
    
    SPX: function (motherboard, self) {
        self.registries["x"] = self.registries["x"] > self.registries["temp"];
        return (0x00);
    },

    SPY: function (motherboard, self) {
        self.registries["y"] = self.registries["y"] > self.registries["temp"];
        return (0x00);
    },

    INX: function (motherboard, self) {
        self.registries["x"] = self.registries["x"] < self.registries["temp"];
        return (0x00);
    },

    INY: function (motherboard, self) {
        self.registries["y"] = self.registries["y"] < self.registries["temp"];
        return (0x00);
    },

    IEX: function (motherboard, self) {
        self.registries["x"] = self.registries["x"] <= self.registries["temp"];
        return (0x00);
    },

    IEY: function (motherboard, self) {
        self.registries["y"] = self.registries["y"] <= self.registries["temp"];
        return (0x00);
    },

    SEX: function (motherboard, self) {
        self.registries["x"] = self.registries["x"] >= self.registries["temp"];
        return (0x00);
    },

    SEY: function (motherboard, self) {
        self.registries["y"] = self.registries["y"] >= self.registries["temp"];
        return (0x00);
    },

    NOX: function (motherboard, self) {
        self.registries["x"] = ~self.registries["x"];
        return (0x00);
    },

    NOY: function (motherboard, self) {
        self.registries["y"] = ~self.registries["y"];
        return (0x00);
    },

    CNE: function (motherboard, self) {
        if (!self.registries["temp"])
            self.registries["ptr_prg"] += 1;
        return (0x00);
    },

    HALT: function (motherboard, self) {
        self.registries["ptr_prg"] -= 1;
        return (0x00);
    },

    XPP: function (motherboard, self) {
        self.registries["x"] = self.registries["ptr_prg"];
        return (0x00);
    },

    YPP: function (motherboard, self) {
        self.registries["y"] = self.registries["ptr_prg"];
        return (0x00);
    },
    
    SWX: function (motherboard, self) {
        self.registries["x"], self.registries["temp"] = self.registries["temp"], self.registries["x"];
        return (0x00);
    },

    SWY: function (motherboard, self) {
        self.registries["y"], self.registries["temp"] = self.registries["temp"], self.registries["y"];
        return (0x00);
    },

    REX: function (motherboard, self) {
        self.registries["x"] = 0x00;
        return (0x00);
    },

    REY: function (motherboard, self) {
        self.registries["y"] = 0x00;
        return (0x00);
    },

    XMP: function (motherboard, self) {
        self.registries["x"] = self.registries["ptr_mem"];
        return (0x00);
    },

    YMP: function (motherboard, self) {
        self.registries["y"] = self.registries["ptr_mem"];
        return (0x00);
    },

    STX: function (motherboard, self) {
        var dev = get_device(motherboard, self.registries["prg_device"]);

        self.registries["ptr_prg"] += 1;
        if (dev == undefined)
            return (0x01);
            self.registries["x"] = dev.read(self.registries["ptr_prg"]);
        return (0x00);
    },

    STY: function (motherboard, self) {
        var dev = get_device(motherboard, self.registries["prg_device"]);

        self.registries["ptr_prg"] += 1;
        if (dev == undefined)
            return (0x01);
            self.registries["y"] = dev.read(self.registries["ptr_prg"]);
        return (0x00);
    },

    LOG: function (motherboard, self) {
        console.log(self.registries["temp"]);
        return (0x00);
    }
}
